bin/commitments_deprecater.rb

```diff
@@ -0,0 +1,190 @@
+# frozen_string_literal: true
+
+require "cluster"
+require "services"
+require "deprecation_record"
+
+Services.mongo!
+
+# Takes (shared print) deprecation records, and tries to match them to commitments
+# so that those commitments can be deprecated.
+# Usage: CommitmentsDeprecater.new.run(<list_of_files>)
+
+class CommitmentsDeprecater
+  attr_reader :err, :log_path, :report
+
+  def initialize(verbose: false)
+    @verbose = verbose
+    @header_spec = ["organization", "ocn", "local_id", "deprecation_status"]
+  end
+
+  # Process a list of files.
+  def run(args)
+    args.each do |arg|
+      if File.file?(arg)
+        from_file(arg)
+      else
+        log "arg #{arg} not a file, skipping"
+      end
+    end
+  end
+
+  # Write log message to file, and STDERR if @verbose
+  def log(msg)
+    if @report.nil?
+      # To reduce the risk of accidental overwrites by a different instance,
+      # make the filename rather unique-ish.
+      iso_stamp = Time.now.strftime("%Y%m%d-%H%M%S")
+      rand_str  = SecureRandom.hex(8)
+      log_dir   = Settings.deprecation_report_path
+      FileUtils.mkdir_p(log_dir)
+      @log_path = "#{log_dir}/commitments_deprecater_#{iso_stamp}_#{rand_str}.log"
+      warn "Logging to #{@log_path}" if @verbose
+      @report = File.open(@log_path, "w")
+    end
+    @report.puts msg
+    warn msg if @verbose
+  end
+
+  # Read deprecation records from file and try to deprecate the matching commitments.
+  def from_file(path)
+    inf = File.open(path, "r")
+    log "Reading deprecation records from #{path}"
+    if !check_header(inf.gets.strip)
+      log "Stopped processing file #{inf} due to header error."
+      return
+    end
+    inf.each_line do |line|
+      success = try_deprecate(line.strip)
+      log "Could not deprecate record" unless success
+    end
+  end
+
+  # Deprecation files must have this header.
+  def check_header(header_line)
+    expected = @header_spec.join("\t")
+    log "Checking header line #{header_line}"
+    if header_line != expected
+      @err << [
+        "Header not OK.",
+        "Observed format:\t#{header_line}",
+        "Expected format:\t#{expected}"
+      ].join("\n")
+      return false
+    end
+
+    log "Header OK"
+    return true
+  end
+
+  # Do what you can with a line, store errors in @err
+  # return true if success / false otherwise
+  def try_deprecate(line)
+    clear_err
+    dep = DeprecationRecord.parse_line(line)
+    log "Find and deprecate a commitment based on the deprecation record: <#{dep}>."
+
+    # dep must match a cluster (on ocn)
+    cluster          = check_cluster(dep.ocn)
```

Comment:
As I understand it essentially if it fails the condition the remaining operations are no-ops. It might be easier to follow the logic if we structured this essentially as a set of guard conditions. I also wonder if it would make sense to move this method (and the `check_*` methods it calls) to `DeprecationRecord`... more thoughts/examples later

Reply:
Looking through this a bit more, from a _logic_ perspective we could simplify this quite a bit to get straight to the non-deprecated commitments for this organization with matching local IDs, but we want to have error logging that helps debug what the issue is. I might be tempted still to separate out the error reporting logic from the whittling-down logic, at least for the steps that are fast.

To be honest, I've always struggled with error reporting like this -- ingest suffers a lot of from this as well; the goals of trying to achieve clear logic, have functions with a single responsibility, and at the same time avoid repeating work are often in tension. I think it would be worth spending some time pairing on this and seeing if we can refactor it some, especially if we can do so without adding a bunch of extra complexity and abstraction around validations or some such.

Reply:
Right. You'll notice that the `Cluster.where()` query just looks for an OCN. A lot could have been done here, but I wanted to be able to log along every step.